rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,675 @@
+   
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7
+```
+
+```console
+nix --supported-language-versions
+6
+7
+```
+
+## Version interoperability
+
+### Versioned built-ins can be passed across file boundaries
+
+```nix
+# a.nix
+version 6;
+builtins.langVersion
+```
+
+```nix
+# b.nix
+version 7;
+[ (import ./a.6.nix) builtins.langVersion ]
+```
+
+```console
+$ nix-instantiate --eval b.nix
+[ 6 7 ]
+```
+
+### Expressions are not forward compatible
+
+```nix
+# a.nix
+version 6;
+import ./b.nix
+```
+
+```nix
+# b.nix
+version 7;
+builtins.langVersion
+```
+
+```console
+$ nix-instantiate --eval a.nix
+error: unsupported Nix language version 7
+```
+
+
+### Best-effort interoperability
+
+```nix
+# a.nix
+version 6;
+{ increment }:
+increment 1
+```
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { increment = x: x + 1 }
+```
+
+Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:
+
+```console
+$ nix-instantiate --eval b.nix
+2
+```
+
+### Pathological example
+
+Usually existing code will be interacted with by calling functions.
+When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.
+
+```nix
+# a.nix
+version 6;
+{ value }:
+value + 1
+```
+
+Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { value = %5 + 7i%; }
+```
+
+```console
+$ nix-instantiate --eval bnix
+error: unsupported value type `complex` at built-in operator `+`
+```
+
+## Deprecation warnings
+
+```
+nix eval --json ./test.nix
+warning: URL literals are deprecated (url-literal)
+         please replace this with a string: "https://nixos.org"
+
+       at test.nix:1:1:
+
+            1| https://nixos.org
+             | ^
+
+"https://nixos.org"
+
+warning: The following deprecated features were used:
+  - url-literal (httsp://nixos.org), 1 time
+
+  Add `--lang-warn-verbose` to show all occurrences
+  Use `--lang-no-warn=url-literal` to disable this warning.
+  Use `--lang-error=url-literal` to issue errors instead of warnings.
+```
+
+## Drawbacks
```

Comment:
Not sure if you should call it a drawback or a risk, but there have been mistakes in implementing the existing preferences about backwards compatibility, thus it is unclear if changes will always be properly evaluated in the future. And false promises are worse than vague promises.

Reply:
People make mistakes, and one cannot commit to outcomes. But one can commit to process. This proposal is to define a process.

Reply:
Every RFC implementation risks implementation bugs. I don't think it needs to be mentioned.

But maybe @7c6f434c is trying to highlight that supporting multiple versions concurrently adds complexity to the project and increases maintenance burden and general likelihood of bugs down the road? I think that is fair to note as this may be more relevant than in other RFCs as this is a requirement that will affect future changes, not just a one time implementation.

Reply:
Added this as a risk: https://github.com/NixOS/rfcs/pull/137/files#r1230919854

Reply:
> Every RFC implementation risks implementation bugs. I don't think it needs to be mentioned.

Not all implementation bugs are created equal. 

Some RFCs allow for a one-time implementation. There will be bugs, they can be found and fixed, and then the solution evolves leaving the RFC behind as an important part of the history. Some RFCs declare decision-making processes that are clearly useful even with some errors on top, and if an error is noticed then the deliberations can be corrected and eventually the decision is made.

The point here is to define compatibility guarantees, and we know there will be higher error rate than in some of the projects used as points of reference, so the question of what error rate is still useful, and how to handle violations merit more detailed consideration.